all-sdk/autotuning.md

```diff
@@ -0,0 +1,477 @@
+}
+
+impl WorkerConfig {
+  /// Set the slot supplier for workflow tasks, using the default fixed-size workflow cache
+  pub fn with_workflow_slot_supplier(&mut self, supplier: impl WorkflowTaskSlotSupplier) -> &mut Self { /* ... */ }
+}
+```
+
+This is analogous to what we'd expect in the worker config for other languages.
+
+The suppliers are kept inside of `Arc`s so that they can be shared across multiple workers, and used with the pauseable
+supplier (which the user presumably wants to retain a reference to so they can call pause).
+
+Of course, most users aren't going to be implementing this in Rust, they'll be doing it in their language, and then
+we've got to translate that into calls across the language barrier. As such we need a slot provider implementation
+that exists in each language's bridge, and calls callbacks into the user's implementation.
+
+Something like:
+
+```rust
+pub struct RustToLangBridgeSlotSupplier {
+  // ... store callbacks etc
+}
+
+impl<T: SlotKind> SlotSupplier<SlotKind=T> for RustToLangBridgeSlotSupplier {
+  // ... call appropriate callbacks, convert data types, etc
+}
+```
+
+## A neat example
+
+For fun, here's a neat example of some of the complex stuff you could get up to with a custom slot supplier.
+
+Say you have some activity workers who all interact with an external service. There are two types of activities that
+interact with the service, one of which can have many instances run in parallel, and another that should prevent any
+new work from being done until it's finished.
+
+You can implement a slot supplier for these workers that coordinates their efforts! Pseudocode below. This toy
+isn't meant to be taken seriously and is likely prone to a race, but it shows that cross-worker coordination is possible
+and also raises some questions (see later section for more on that).
+
+```rust
+struct MyCoordinatingSlotSupplier {
+  lock_svc_client: LockServiceClient,
+  per_worker_allowed: Sempahore,
+}
+
+impl ActivityTaskSlotSupplier for MyCoordinatingSlotSupplier {
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error> {
+    self.lock_svc_client.wait_for_lock_not_held_or_being_acquired().await?;
+    self.per_worker_allowed.acquire().await?;
+    Ok(())
+  }
+  
+  fn try_reserve_slot(&self) -> Result<bool, anyhow::Error> { /* ... */ }
+  
+  fn mark_slot_used(&self, info: &ActivitySlotInfo) {
+    if info.activity_type == "there_can_only_be_one" {
+      // Probably kicks off something in the background, but immediately blocks slot reservation
+      self.lock_svc_client.eventually_acquire_lock();
+    }
+  }
+  
+  fn release_slot(&self, info: &ActivitySlotInfo) {
+    if info.activity_type == "there_can_only_be_one" {
+      self.lock_svc_client.release_lock();
+    }
+    self.per_worker_allowed.release();
+  }
+}
+```
+
+## Drawbacks
+
+* Gives users a big footgun. This is mitigated by the fact that most users are probably just going to use one of our
```

Comment:
Anything we can do to mitigate? For example, letting users fallback to a default implementation if something goes wrong?

Reply:
Not really possible - there's not any way to distinguish between the user's implementation intentionally not handing out slots, or it not handing out slots because it's broken.